docs/contributing: Initial Retries and Waiters documentation (#17508)
hashicorp/terraform-provider-aws: docs/contributing/retries-and-waiters.md

# Retries and Waiters

_Please Note: This documentation is intended for Terraform AWS Provider code developers. Typical operators writing and applying Terraform configurations do not need to read or understand this material._

> Terraform plugins may run into situations where repeatedly checking or calling the remote system after an operation may be necessary. These typically fall under three classes. The first where the operation never reaches the remote system, the second where the operation reaches the remote system and the remote system responds that it temporarily cannot handle the request, and third where the implementation of the remote system requires additional operations to ensure success. This guide is designed to describe each of these situations and highlight implementations that can ensure operation success.

Author: ```suggestion
Terraform plugins may run into situations where calling the remote system after an operation may be necessary. These typically fall under three classes where:

- The request never reaches the remote system.
- The request reaches the remote system and responds that it cannot handle the request temporarily.
- The implementation of the remote system requires additional requests to ensure success.

This guide describes the behavior of the Terraform AWS Provider and provides code implementations that help ensure success in each of these situations.
```

### State Change Configuration and Functions

The [`resource.StateChangeConf` type](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#StateChangeConf) along with its receiver methods [`WaitForState()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#StateChangeConf.WaitForState) and [`WaitForStateContext()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#StateChangeConf.WaitForStateContext) is a generic primitive for repeating operations in Terraform resource logic until desired value(s) are received. The "state change" in this case is generic to any value and not specific to the Terraform State. Among other functionality, it supports some of these desirable optional properties:

> - Expected pending value(s) while waiting for the target value(s) to be reached. Unexpected values are returned as an error which can be augmented with additional details.

Author: Avoiding the list starter of "optional" then immediately using "requiring"

```suggestion
- Expecting specific value(s) while waiting for the target value(s) to be reached. Unexpected values are returned as an error which can be augmented with additional details.
```

> - Requiring the target value(s) to be returned multiple times in succession.

Author: ```suggestion
- Expecting the target value(s) to be returned multiple times in succession.
```

_NOTE: The section descibes the current handling with version 1 of the AWS Go SDK. In the future, this codebase will be migrated to version 2 of the AWS Go SDK. The newer version implements additional retry conditions by default, such as consistently retrying all common network errors._

_NOTE: The section descibes the current handling with Terraform Plugin SDK resource signatures without `context.Context`. In the future, this codebase will be migrated to the context-aware resource signatures which currently enforce a 20 minute default timeout that conflicts with the timeout with the default `max_retries` value. The Terraform Plugin SDK may be updated to support removing this default 20 minute timeout or the default retry mechanism described here will be updated to prevent context cancellation errors where possible._

### Lower Network Error Retries

> Given the very high default number of AWS Go SDK retries and the associated timeout before operators would receive the error, the [`hashicorp/aws-sdk-go-base` codebase](https://github.com/hashicorp/aws-sdk-go-base/blob/57529b4c2d2f8f3b5299d66a829b01259fa800d7/session.go#L108-L130) enforces lower retries (10) for certain network errors that typically cannot be remediated via retries. This roughly equates to 30 seconds of retries.

Author: I meant to clarify that due to _our_ very high default configuration.

```suggestion
Given the very high default number of AWS Go SDK retries configured in the Terraform AWS Provider and the excessive wait that practitioners would face, the [`hashicorp/aws-sdk-go-base` codebase](https://github.com/hashicorp/aws-sdk-go-base/blob/57529b4c2d2f8f3b5299d66a829b01259fa800d7/session.go#L108-L130) lowers retries to 10 for certain network errors that typically cannot be remediated via retries. This roughly equates to 30 seconds of retries.
```

### Terraform AWS Provider Service Retries

The AWS Go SDK middleware provides hooks for injecting custom logic in the service client handlers. This type of handling is preferred when retry behavior must be applied in many resources, such as an error code that is not marked as automatically retryable by the service API. The Terraform AWS Provider includes some additional retry changing behaviors using this method, which can be found in the `aws/config.go` file. For example:

```go
client.kafkaconn.Handlers.Retry.PushBack(func(r *request.Request) {
	if tfawserr.ErrMessageContains(r.Error, kafka.ErrCodeTooManyRequestsException, "Too Many Requests") {
		r.Retryable = aws.Bool(true)
	}
})
```

## Eventual Consistency

> Eventual consistency is a temporary condition where the remote system can return outdated information or errors due to not being strongly read-after-write consistent. This is a pattern found in remote systems that must be highly scaled for broad usage.

Author: Let's followup with rewriting this section after merge.

Terraform expects any planned resource lifecycle change (create, update, destroy of the resource itself) and planned resource attribute value change to match after being applied. Conversely, operators typically expect that Terraform resources also implement the concept of drift detection for resources and their attributes, which requires reading information back from the remote system after an operation. A common implementation is refreshing the Terraform State information (`d.Set()`) during the `Read` function of a resource after `Create` and `Update`.

> These two concepts conflict with each other and require additional handling in Terraform resource logic as shown in the following sections. These issues are _not_ reliably reproducible, especially in the case of writing acceptance testing, so they can be elusive with false positives to verify fixes.

Author: The caveat here is that Terraform will not report the inconsistency as drift unless the configuration happens to be planned again in the eventual consistency window. Instead, Terraform returns errors because the planned value does not match the applied value.

	}

	if err != nil {
		return fmt.Errorf("... error message context ... : %w", err)
	}
```

_NOTE: The section descibes the current handling with version 1 of the AWS Go SDK. In the future, this codebase will be migrated to version 2 of the AWS Go SDK. The newer version natively supports operation-specific retries in a more friendly manner, which may replace this type of implementation._

#### IAM Error Retries

A common eventual consistency issue is an error returned due to IAM permissions. The IAM service itself is eventually consistent along with the propagation of its components and permissions to other AWS services. For example if the following operations occur in quick succession:

- Create an IAM Role
- Attach an IAM Policy to the IAM Role
- Reference the new IAM Role in another AWS service, such as creating a Lambda Function

The last operation can receive varied API errors ranging from:

- IAM Role being reported as not existing
- IAM Role being reported as not having permissions for the other service to use it (assume role permissions)
- IAM Role being reported as not having sufficient permissions (inline or attached role permissions)

> Each AWS service API (and sometimes even operations within the same API) varies in the implementation of these errors. To handle them, it is recommended to use the Operation Specific Error Retries pattern. The Terraform AWS Provider implements a standard timeout constant of two minutes in the `aws/internal/service/iam/waiter` package which should be used for all retry timeouts associated with IAM errors. This timeout was derived from years of Terraform operational experience with all AWS APIs.

Reviewer: Not sure if you want a link here or not as it refers to the parent section.

```suggestion
Each AWS service API (and sometimes even operations within the same API) varies in the implementation of these errors. To handle them, it is recommended to use the [Operation Specific Error Retries](#operation-specific-error-retries) pattern. The Terraform AWS Provider implements a standard timeout constant of two minutes in the `aws/internal/service/iam/waiter` package which should be used for all retry timeouts associated with IAM errors. This timeout was derived from years of Terraform operational experience with all AWS APIs.
```

	// ...

	d.HasChange("attribute") {
		// ... AWS Go SDK logic to update attribute ...

		if err := waiter.ThingAttributeUpdated(conn, d.Id(), d.Get("attribute").(string)); err != nil {
			return fmt.Errorf("error waiting for Example Thing (%s) attribute update: %w", d.Id(), err)
		}
	}

	// ...
}
```

## Asynchronous Operations

Many longer running operations are initiated and performed asynchronously in the background after returning a successful response to the initial operation request. These operations may either have a component level status field that can be tracked (e.g. `CREATING`, `UPDATING`, etc.) or they may have an explicit tracking identifier to follow the progress of the background operation.

Terraform resources are expected to wait for these background operations to complete. Failing to do so may introduce incomplete state information or downstream errors in other resources. In rare scenarios involving very long operations, operators may request a flag to skip the waiting, however these should only be implemented case-by-case to prevent those previously mentioned confusing issues.

### AWS Go SDK Waiters

> In limited cases, the AWS service API model includes the information to automatically generate a waiter function in the AWS Go SDK for an operation. These are typically named with the prefix `WaitUntil...`. If available, these functions are often good enough for an initial resource implementation and can be used. For example:

Author: The term "model" here is important. These are not part of the API itself. Also I think distinctively saying limited (as they are rare) versus some is also important since they are hard to find.

```suggestion
In limited cases, the AWS service API model includes the information to automatically generate a waiter function in the AWS Go SDK for an operation. These are typically named with the prefix `WaitUntil...`. If available, these functions can be used for an initial resource implementation. For example:
```

```go
if err := conn.WaitUntilEndpointInService(input); err != nil {
	return fmt.Errorf("error waiting for Example Thing (%s) ...: %w", d.Id(), err)
}
```

> If it is necessary to customize the timeouts and polling with these, it is generally preferable to use [Resource Lifecycle Waiters](#resource-lifecycle-waiters) instead, since they are more commonly used throughout the codebase.

Reviewer 2: ```suggestion
If it is necessary to customize the timeouts and polling, we generally prefer using [Resource Lifecycle Waiters](#resource-lifecycle-waiters) instead since they are more commonly used throughout the codebase.
```

### Operation Specific Error Retries

Given a properly ordered Terraform configuration, eventual consistency can unexpectedly prevent downstream operations from being successful. To reduce frustrating behavior for operators where a simple retry after a few seconds would resolve the error, AWS Go SDK operations can be wrapped with the `resource.Retry()` or `resource.RetryContext()` function. These retries should have a reasonably low timeout (typically two minutes but up to five minutes) that are saved in a constant for reusability. These functions are preferably inline with the associated resource logic, to remove any indirection with the code.

This type of logic should not be implemented to overcome improperly ordered Terraform configurations as it is not guaranteed to work in larger environments.

```go
// aws/internal/service/{SERVICE}/waiter/consts.go (created if does not exist)
package {SERVICE}

const (
	// Maximum amount of time to wait for Example eventual consistency

> 	ExampleTimeout = 2 * time.Minute

Reviewer: If _Example_ is the service name and we are in a service-specific package, should this be `PropagationTimeout = 2 * time.Minute` instead (like it is for IAM)?

Author: Fixed in latest commit.

When encountering this type of situation, every operation after the initial operation must account for the condition, which can be tricky depending on the resource logic itself. Most resources should be written where the `Create` function will return the `Read` function to fill in all computed attributes and ensure the configuration was applied correctly. In these cases, the `Read` function will need additional logic to overcome the temporary condition on resource creation. For eventually consistent resources, these "not found" type errors can still occur in the `Read` function even after implementing [Resource Lifecycle Waiters](#resource-lifecycle-waiters) in the Create function for the resource itself.

```go
// aws/internal/service/{SERVICE}/waiter/consts.go (created if does not exist)

package {SERVICE}

const (
	// Maximum amount of time to wait for Example eventual consistency on creation

> 	ExampleCreationTimeout = 2 * time.Minute

Reviewer: If _Example_ is the service name and we are in a service-specific package, should this be `CreationTimeout = 2 * time.Minute` instead?

Author: Fixed in latest commit.

### Resource Lifecycle Waiters

Most of the codebase uses `resource.StateChangeConf` and `resource.RefreshStateFunc` handlers for tracking either component level status fields or explicit tracking identifiers. These should be placed in the `aws/internal/service/{SERVICE}/waiter` package and split into separate functions. For example:

```go
// aws/internal/service/{SERVICE}/waiter/status.go (created if does not exist)

// ThingStatus fetches the Thing and its Status
func ThingStatus(conn *example.Example, id string) resource.StateRefreshFunc {
	return func() (interface{}, string, error) {
		output, err := /* ... AWS Go SDK operation to fetch resource/status ... */

		if tfawserr.ErrCodeEquals(err, example.ErrCodeResourceNotFoundException) {
			return nil, "", nil
		}

		if err != nil {
			return nil, "", err
		}

		if output == nil {
			return nil, "", nil
		}

		return output, aws.StringValue(output.Status), nil
	}
}
```

```go
// aws/internal/service/{SERVICE}/waiter/waiter.go (created if does not exist)

const (
	ThingCreationTimeout = 2 * time.Minute

> 	ThingDeletionTimeout = 5 * time.Minute

Reviewer: In the _Operation Specific Error Retries_ and _Resource Lifecycle Retries_ sections above timeout constants are in `aws/internal/service/{SERVICE}/waiter/consts.go`. Is it confusing to potentially have similar constants in two locations?

Author: Fixed in latest commit.
